src/bytes.rs

```diff
@@ -104,16 +105,51 @@ pub struct Bytes {
     data: AtomicPtr<()>,
     vtable: &'static Vtable,
 }
+/// A trait for underlying implementations for `Bytes` type.
+///
+/// All implementations must fulfill the following requirements:
+/// - They are cheaply cloneable and thereby shareable between an unlimited amount
+///   of components, for example by modifying a reference count.
+/// - Instances can be sliced to refer to a subset of the the original buffer.
+pub unsafe trait BytesImpl: 'static {
+    /// Decompose `Self` into parts used by `Bytes`.
+    fn into_bytes_parts(this: Self) -> (AtomicPtr<()>, *const u8, usize);
+
+    /// Creates itself directly from the raw bytes parts decomposed with `into_bytes_parts`.
+    unsafe fn from_bytes_parts(data: &mut AtomicPtr<()>, ptr: *const u8, len: usize) -> Self;
+
+    /// Returns new `Bytes` based on the current parts.
+    unsafe fn clone(data: &AtomicPtr<()>, ptr: *const u8, len: usize) -> Bytes;
```

Comment:
@HyeonuPark - Is there a case in which one would need to clone bytes using the components?  
If so, is there another way to go about this?
A third party container implementor (say,  Mmap https://docs.rs/memmap/latest/memmap/struct.Mmap.html)  isn't going to be able to construct a Bytes instance because they won't have access to the internals of `Bytes`

Reply:
With the current design cloning bytes should use the component, since cloning mechanisms like reference counting are implemented as component. For example they should do things like `Arc::increment_strong_count()` within it. Currently the `Bytes` is shipped with number of builtin components:

- Static component backed by `&'static [u8]` and does nothing on cloning
- Shared component which increase refcount on cloning.
- Promotable component which initially store bytes without `Arc` and be promoted to shared component on first cloning.

And the `BytesMut` has number of components similar but slightly different stretagies. And the promotable one is the only reason we have `&AtomicPtr<()>` not `*const ()`. Is it necessary? I'm not sure, but it was one of the goals of this PR to not change existing behaviors.

For the mmap part, I built my own buffer type from aligned slices of giant single mmap region. First few bytes of each slices are considered header and one of its fields are used as refcount since I don't want additional allocation for it. This refcount is modified on clone and drop.

Reply:
I was referring to the fact that `clone` returns `Bytes`.   A 3rd party implementation isn't going to be able to construct Bytes using the parts, because they don't have access to the internals of `Bytes`.   

We would have to provide some sort of public  `Bytes::from_parts`     but I think that isn't the best approach.  

Better would be that the trait doesn't specify that the implementation returns `Bytes`.  Instead, the consumer of the clone function (a Bytes instance) would receive the new parts, and know how to construct them.